src/core/vector/qgsvectorlayerutils.cpp

```diff
@@ -120,6 +120,52 @@ QList<QVariant> QgsVectorLayerUtils::getValues( const QgsVectorLayer *layer, con
   return values;
 }
 
+QList<QVariant> QgsVectorLayerUtils::uniqueValues( const QgsVectorLayer *layer, const QString &fieldOrExpression, bool &ok, bool selectedOnly, int limit, QgsFeedback *feedback )
+{
+  QSet<QVariant> uniqueValues;
+  QgsFeatureIterator fit = getValuesIterator( layer, fieldOrExpression, ok, selectedOnly );
+  if ( ok )
+  {
+    std::unique_ptr<QgsExpression> expression;
+    QgsExpressionContext context;
+
+    int attrNum = layer->fields().lookupField( fieldOrExpression );
+    if ( attrNum == -1 )
+    {
+      // use expression, already validated in the getValuesIterator() function
+      expression.reset( new QgsExpression( fieldOrExpression ) );
+      context.appendScopes( QgsExpressionContextUtils::globalProjectLayerScopes( layer ) );
+    }
```

Comment:
This new function is considerably less optimised then the existing QgsCategorizedSymbolRendererWidget::layerUniqueValues implementation for non-expression based values. The existing one delegates to `mLayer->uniqueValues`, which in turn directly calls `mDataProvider->uniqueValues`. The data provider version is heavily optimised, because eg in the case of postgres provider it gets the database to calculate the values instead of iterating on the QGIS side.

Can you please re-introduce this optimisation in the cases where it's safe to do so?

Reply:
Thanks for the explanation. Indeed, one can use `QgsVectorLayer::uniqueValues` in the non expression case when `selectedOnly` is `False`.

Done so.